### DISCUSSION TITLE
Qelm V-0.1.4 Release

### DISCUSSION THREAD

**Discussion by Inserian** [R-D-BioTech-Alaska/Qelm#12] [Announcements]
Qelm upgrades, changes and tweaks.

I have been thoroughly testing each section broken apart into small single units by creating testing applications for each section. The following upgrades will be in the newest version releasing in the next week.

If you would like to see any additions, tweaks or upgrades, feel free to submit discussions or projects to us.
---

### 1. GUI Initialization & Thread Safety

* **Theme fallback**

  ```python
  try:
      style.theme_use('clam')
  except:
      logging.warning("…")
  ```

  prevents `sys.exit()` on missing theme.
* **Main‑thread UI updates**
  All calls to `messagebox` or widget state changes from background threads now use:

  ```python
  self.master.after(0, lambda: …)
  ```

---

### 2. Reusable Qubit Pool & Reset Loop

* **`QuantumChannelManager`**

  * Manages a pool of `QuantumChannel` objects.
  * `allocate_channels(n)` / `release_channels(list)` replaces ad‑hoc `QuantumCircuit(...)` recreation.
* **Proper mid‑circuit measure + reset**

  ```python
  qc.measure(…)
  qc.reset(…)
  ```

  ensures each physical qubit is recycled slice‑to‑slice without cloning.

---

### 3. Sub‑Bit Encoding/Decoding

* **New methods** on `QuantumChannel`:

  ```python
  encode_subbit((θ, φ))
  decode_subbit()  # returns θ, φ
  ```
* **`qc.use_subbit` flag**
  toggles between standard amplitude encoding and θ/φ “phase‑band” sub‑bit mode.

---

### 4. Grover Search Fully Implemented

* **`GroverOracle`**, **`GroverDiffuser`**, **`GroverSearch`** classes no longer stubs—now produce real amplitude amplification and return counts via `.run()`.

---

### 5. Basic Error‑Mitigation Hook

* **`entropy_factor`** on each channel applies tiny random Ry rotations for Pauli‑twirling.
* **Teleport‑reset** and ZNE/CDR insertion points scaffolded in the slice loop.

---

### 6. Modular Quantum Layers

* Consolidated simulator selection in `QuantumLayerBase.initialize_simulator()`.
* Split old monolithic attention/FFN into:

  * `QuantumAttentionLayer`
  * `QuantumFeedForwardLayer`
    each with `build_simple_circuit()` / `build_advanced_circuit()`.

---

### 7. Parallel Parameter‑Shift Gradients

* **`compute_gradients_parallel()`** uses `ProcessPoolExecutor` for full‑vector gradient evaluation.
* Progress updates every 100 parameters via a `progress_callback` feeding the GUI’s gradient bar.

---

### 8. Data & Tokenization Pipeline

* **`ExponentialSubwordTokenizer`** unchanged but now seamlessly feeds into the new channel‑pool loop.
* Clear branches for **real** vs. **synthetic** dataset loading—no hidden defaults.

---

### 9. Logging & Resource Monitoring

* **Elapsed / Remaining time** computed in the training loop and shown live should work more accurately now.

---

### 10. Serialization & Versioning (I'm testing this to see if it will be more reliable for good models, it can currently cause issues with old models.)

* Added `"version": "4.0"` to all saved JSON models for compatibility checks.
* Combined model + token‑map save/load into `save_model_and_tokens()` / `load_model_and_tokens()`, with strict size verification.

---

If you continue to experience issues with error mitigation, or auto close please go to the start a new issue section or respond here.

**Comment by Inserian** [comment]
Also, if you want to change the critical aspect of qubit spin optimization which I moronically never wrote into the gui (trust me, tkinkter is a buzzkill). Alter the optimization level -
```
def measure_qubit_spin_z(qc: "QuantumChannel") -> float:
    temp_circuit = qc.circuit.copy()
    backend = qc.backend
    

> optimized_circuit = transpile(temp_circuit, backend, optimization_level=3) <---- Fix this (2 should be safe but if you want speed 0 or 1)

    ensure_single_statevector(optimized_circuit)
    job = backend.run(optimized_circuit)
    result = job.result()
    statevector_obj = result.get_statevector(optimized_circuit)
    statevector = np.asarray(statevector_obj)
    alpha = np.abs(statevector[0])**2
    beta = np.abs(statevector[1])**2
    return round(alpha - beta, 4)

**Comment by Inserian** [comment]
Figured out some new methods for encode and decode using my built in cubit, using this method I have been able to shave 85% training time per gradient in the newest model.
![optimized](https://github.com/user-attachments/assets/171e19cd-5385-4013-9bd4-75667e739b9e)
